x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardChangesAction.java

```diff
@@ -265,19 +282,58 @@ public TransportAction(Settings settings,
             final long mappingVersion = clusterService.state().metaData().index(shardId.getIndex()).getMappingVersion();
 
             final Translog.Operation[] operations = getOperations(
                     indexShard,
                     seqNoStats.getGlobalCheckpoint(),
-                    request.fromSeqNo,
-                    request.maxOperationCount,
-                    request.expectedHistoryUUID,
-                    request.maxOperationSizeInBytes);
-            return new Response(mappingVersion, seqNoStats.getGlobalCheckpoint(), seqNoStats.getMaxSeqNo(), operations);
+                    request.getFromSeqNo(),
+                    request.getMaxOperationCount(),
+                    request.getExpectedHistoryUUID(),
+                    request.getMaxOperationSizeInBytes());
+            return getResponse(mappingVersion, seqNoStats, operations);
+        }
+
+        @Override
+        protected void asyncShardOperation(
+                final Request request,
+                final ShardId shardId,
+                final ActionListener<Response> listener) throws IOException {
+            final IndexService indexService = indicesService.indexServiceSafe(request.getShard().getIndex());
+            final IndexShard indexShard = indexService.getShard(request.getShard().id());
+            final SeqNoStats seqNoStats = indexShard.seqNoStats();
+
+
+            if (request.getFromSeqNo() > seqNoStats.getGlobalCheckpoint()) {
+                assert request.getFromSeqNo() == 1 + seqNoStats.getGlobalCheckpoint();
+                indexShard.addGlobalCheckpointListener(
+                        request.getFromSeqNo(),
+                        (g, e) -> {
+                            if (g == UNASSIGNED_SEQ_NO) {
+                                assert e != null;
+                                if (e instanceof TimeoutException) {
+                                    final long mappingVersion =
+                                            clusterService.state().metaData().index(shardId.getIndex()).getMappingVersion();
+                                    final SeqNoStats latestSeqNoStats = indexShard.seqNoStats();
+                                    listener.onResponse(getResponse(mappingVersion, latestSeqNoStats, EMPTY_OPERATIONS_ARRAY));
+                                } else {
+                                    listener.onFailure(e);
+                                }
+                            } else {
+                                try {
+                                    super.asyncShardOperation(request, shardId, listener);
+                                } catch (final IOException e1) {
+                                    listener.onFailure(e1);
+                                }
+                            }
+                        },
+                        request.getPollTimeout());
+            } else {
+                listener.onResponse(shardOperation(request, shardId));
```

Comment:
do we need to fork on the get threadpool here? This what `TransportSingleShardAction#asyncShardOperation()` does.

Reply:
I pushed d4f8a9818525579bdac3a05083a81cf2b5ebc996.